File: workspace/src/my_py_pkg/my_py_pkg/publisher_with_params.py
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from std_msgs.msg import String

class PublisherWithParams(Node):
    """Publisher example that periodically sends out a string"""

    def __init__(self):
        """Constructor"""

        # Call the Node class constructor with the node name
        super().__init__('publisher_with_params')

        # Declare parameters
        self.declare_parameter('message', "Hello")
        self.declare_parameter('timer_period', 1.0)

        # Set to attributes
        self._message = self.get_parameter('message').value
        self._timer_period = self.get_parameter('timer_period').value

        # Configure callback for runtime parameter update
        self.add_post_set_parameters_callback(self._post_parameters_callback)

        # Create a publisher object
        self._publisher = self.create_publisher(String, 'my_topic', 10)

        # Periodically call method
        self._timer = self.create_timer(
            self._timer_period, 
            self._timer_callback
        )

    def _timer_callback(self):
        """Publishes simple message to topic"""

        # Fill out String message
        msg = String()
        msg.data = self._message

        # Publish message to topic
        self._publisher.publish(msg)
        self.get_logger().info(f"Publishing: {msg.data}")

    def _post_parameters_callback(self, params):
        """Set parameter after node started"""

        # Loop through all parameters
        for param in params:

            # Update message
            if param.name == 'message':
                self._message = param.value
                self.get_logger().info(f"Set {param.name} to {param.value}")

            # Update timer
            elif param.name == 'timer_period':

                # Update member variable
                self._timer_period = param.value
                self.get_logger().info(f"Set {param.name} to {param.value}")

                # Reset timer
                self._timer.cancel()
                self._timer = self.create_timer(
                    self._timer_period, 
                    self._timer_callback
                )

            # Unknown parameter
            else:
                self.get_logger().warn(f"Unknown parameter: {param.name}")

def main(args=None):
    """Main entrypoint"""

    # Initialize and run node
    try:
        rclpy.init()
        node = PublisherWithParams()
        rclpy.spin(node)

    # Catch ctrl+c or shutdown request
    except (KeyboardInterrupt, ExternalShutdownException):
        pass

    # Destroy node (now) and gracefully exit
    finally:
        if node is not None:
            node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()

File: workspace/src/my_cpp_pkg/src/publisher_with_params.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

/**
 * Publisher example that periodically sends out a string
 */
class PublisherWithParams : public rclcpp::Node
{

public:

  /**
   * Constructor (call Node class constructor with node name)
   */
  PublisherWithParams() : Node("publisher_with_params")
  {
    // Get parameters
    this->declare_parameter("message", "Hello");
    this->declare_parameter("timer_period", 1.0);

    // Set parameters to member variables
    message_ = this->get_parameter("message").as_string();
    timer_period_ = this->get_parameter("timer_period").as_double();

    // Configure callback for runtime parameter update
    param_cb_handle_ = this->add_post_set_parameters_callback(
      std::bind(
        &PublisherWithParams::post_parameters_callback, 
        this, 
        std::placeholders::_1));

    // Create a publisher object
    publisher_ = this->create_publisher<example_interfaces::msg::String>(
      "my_topic", 10);

    // Periodically call method
    timer_ = this->create_wall_timer(
      std::chrono::duration<double>(timer_period_),
      std::bind(&PublisherWithParams::timer_callback, this));
  }

private:

  /**
   * Publishes simple message to topic
   */
  void timer_callback()
  {
    // Fill out String message
    auto msg = example_interfaces::msg::String();
    msg.data = message_;

    // Publish message to topic
    publisher_->publish(msg);
    RCLCPP_INFO(this->get_logger(), "Publishing: %s", msg.data.c_str());
  }

  /**
   * Set parameter after node started
   */
  void post_parameters_callback(
    const std::vector<rclcpp::Parameter> &parameters)
  {
    //Loop through all parameters
    for (const auto &param : parameters) {

      // Update message
      if (param.get_name() == "message") {
        message_ = param.as_string();
        RCLCPP_INFO(
          this->get_logger(), 
          "Set %s to %s", 
          param.get_name().c_str(), 
          param.as_string().c_str());
      }

      // Update timer period
      else if (param.get_name() == "timer_period") {

        // Update member variable
        timer_period_ = param.as_double();
        RCLCPP_INFO(
          this->get_logger(), 
          "Set %s to %f", 
          param.get_name().c_str(), 
          param.as_double());

        // Reset timer
        timer_->cancel();
        timer_ = this->create_wall_timer(
          std::chrono::duration<double>(timer_period_),
          std::bind(&PublisherWithParams::timer_callback, this));
      }

      // Unknown parameter
      else {
        RCLCPP_WARN(
          this->get_logger(), 
          "Unknown parameter: %s", 
          param.get_name().c_str());
      }
    }
  }

  // Declare member variables
  std::string message_;
  double timer_period_;
  rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::node_interfaces::PostSetParametersCallbackHandle::SharedPtr 
    param_cb_handle_;
};

/**
 * Main entrypoint
 */
int main(int argc, char * argv[])
{
  // Initialize ROS 2
  rclcpp::init(argc, argv);

  // Initialize and run node
  auto node = std::make_shared<PublisherWithParams>();
  rclcpp::spin(node);

  // Cleanup
  rclcpp::shutdown();

  return 0;
}
